cache/cache.go

```diff
@@ -0,0 +1,130 @@
+// Package cache implements a fixed size circular cache which can be used to
+// cache a fixed number of items. Once the underlying array is full, the oldest
+// item in the cache will be overwritten. The cache thread-safe.
+package cache
+
+import (
+	"errors"
+	"sync"
+)
+
+// noIndex is the value of the cache's index when no values have been added.
+const noIndex = -1
+
+// errInvalidSize is returned when an invalid size for a cache is provided.
+var errInvalidSize = errors.New("cache size must be > 0")
+
+// Circular is a cache which retains a set of values in memory, and
+// overwrites the oldest item in the cache when a new item needs to be
+// written.
+type Circular struct {
+	// size is the total number of items that the cache will hold.
+	size int
+
+	// index is the index of the most recently written item in the
+	// cache.
+	index int
+
+	// total is the total number of items that have been added to the
+	// cache.
+	total int
+
+	// items is the set of cached items.
+	items []interface{}
+
+	sync.RWMutex
+}
+
+// NewCircular returns a new circular cache with the size provided. It will fail
+// if a zero of negative size parameter is provided.
+func NewCircular(size int) (*Circular, error) {
+	if size <= 0 {
+		return nil, errInvalidSize
+	}
+
+	return &Circular{
+		index: noIndex,
+		total: 0,
+		size:  size,
+
+		// Create a slice with length and capacity equal to the size of
+		// the cache so that we do not need to resize the underlying
+		// array when we add items.
+		items: make([]interface{}, size),
```

Comment:
Instead of taking up a fixed size, this could grow as well. Not a big thing as for now we only use this with a handful of peers.

Reply:
> make([]interface{}, 0, size)

Thought of going this route but then you end up having custom logic for the first set of elements you add - need to append for the first `size` x elements (so that slice len is increased) and then store at index after that. Since we don't really need `len` for much, I thought it would be better to have more simple add logic?